File: challenges/exercism/go/exercises/acronym/acronym.go
package acronym

import (
	"strings"
	"unicode"
)

// Checks if the input rune is a letter.
func IsLetter(s string, i int) bool {
	cur := rune(s[i])

	if i == 0 { return unicode.IsLetter(cur) }

	prev := rune(s[i - 1])

	return unicode.IsLetter(cur) && !unicode.IsLetter(prev) && prev != '\''
}

// Creates an abbreviation from the input string.
func Abbreviate(s string) string {
	abbr := ""

	for i, r := range s {
		if IsLetter(s, i) {
			abbr += strings.ToUpper(string(r))
		}
	}

	return abbr
}

File: challenges/exercism/c/exercises/acronym/src/acronym.c
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

#include "acronym.h"

#define strlen_s(s) (s == NULL ? 0 : strlen(s))

static int is_first_character_of_word(const char *phrase, int index)
{
	char cur = phrase[index];

	if (index == 0) return isalpha(cur);

	char prev = phrase[index - 1];

	return isalpha(cur) && !isalpha(prev) && prev != '\'';
}

char *abbreviate(const char *phrase)
{
	char *acronym = NULL;

	long phrase_len = strlen_s(phrase);
	long acronym_len = 1;
	long letters_count = 0;

	for (int i = 0; i < phrase_len; i++) {
		if (is_first_character_of_word(phrase, i)) {
			// Check if reallocation is needed
			if (letters_count == acronym_len - 1) {
				acronym = realloc(acronym, acronym_len * 2);
				acronym_len *= 2;
			}

			acronym[letters_count++] = toupper(phrase[i]);
		}
	}

	return acronym;
}
